content/3_Silver/Sorting_Custom.mdx

```diff
@@ -640,6 +639,203 @@ public class Sol {
+discuss how coordinate compression is applied. Since the solution uses
+[2D-prefix](/silver/prefix-sums-2) sums (another Silver topic), it is helpful if
+all point coordinates are coordinate-compressed to the range $0$ to $N-1$ so
+they can be used as array indices.
+
+Below you will find the solution code to Rectangular Pasture, which uses
+coordinate compression at the beginning. Observe how a custom comparator is used to sort the
+points:
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+BeginCodeSnip{}
+
+#include <iostream>
+#include <algorithm>
+using namespace std;
+ 
+typedef pair<int,int> Point;
+bool ycomp(Point p, Point q) { return p.second < q.second; }
+ 
+const int MAX_N = 2500;
+int N, Psum[MAX_N+1][MAX_N+1];
+Point P[MAX_N];
+ 
+int rsum(int x1, int y1, int x2, int y2)
+{
+  return Psum[x2+1][y2+1] - Psum[x2+1][y1] - Psum[x1][y2+1] + Psum[x1][y1];
+}
+ 
+int main(void)
+{
+  cin >> N;
+  for (int i=0; i<N; i++) {
+    int x, y;
+    cin >> x >> y;
+    P[i] = make_pair(x,y);
+  }
+  //------------- (coordinate compression) ----------
+  sort(P, P+N); 
+  for (int i=0; i<N; i++) P[i].first = i+1;
+  sort(P, P+N, ycomp); 
+  for (int i=0; i<N; i++) P[i].second = i+1;
+  //-------------------------------------------------
+  for (int i=0; i<N; i++) Psum[P[i].first][P[i].second] = 1;
+  for (int i=1; i<=N; i++)
+    for (int j=1; j<=N; j++)
+      Psum[i][j] += Psum[i-1][j] + Psum[i][j-1] - Psum[i-1][j-1];
+  long long answer = 0;
+  for (int i=0; i<N; i++) 
+    for (int j=i; j<N; j++) {
+      int x1 = min(P[i].first, P[j].first) - 1;
+      int x2 = max(P[i].first, P[j].first) - 1;
+      answer += rsum(0,i,x1,j) * rsum(x2,i,N-1,j);
+    }
+  cout << answer + 1 << "\n";
+  
+}
+
+
+EndCodeSnip
+
+```
+
+</CPPSection>
+
+<JavaSection>
+
+```java
+
+BeginCodeSnip{}
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.Scanner;
+ 
+public class RectangularPasture {
+    static int[][] sums;
+ 
+    static int getSum(int fromX, int toX, int fromY, int toY) {
+        return sums[toX][toY] - sums[fromX - 1][toY] - sums[toX][fromY - 1] + sums[fromX - 1][fromY - 1];
+    }
+ 
+    public static void main(String[] args) {
+        Scanner in = new Scanner(System.in);
+        int n = in.nextInt();
+        int[] xs = new int[n];
+        int[] ys = new int[n];
+        Integer[] cows = new Integer[n];
+        for (int j = 0; j < n; j++) {
+            xs[j] = in.nextInt();
+            ys[j] = in.nextInt();
+            cows[j] = j;
+        }
+        // ------------- (coordinate compression) ----------
+        Arrays.sort(cows, Comparator.comparingInt(j -> xs[j]));
+        for (int x = 1; x <= n; x++) {
+            xs[cows[x - 1]] = x;
+        }
+        Arrays.sort(cows, Comparator.comparingInt(j -> ys[j]));
+        for (int y = 1; y <= n; y++) {
+            ys[cows[y - 1]] = y;
+        }
+	// -------------------------------------------------
```

Comment:
What i meant is that you can snip out everything except this portion.

Reply:
The point is you are displaying the "important" part and hiding the rest. If someone wants to see the full implementation, they can expand the snips. Otherwise, only the part we want to emphasize should be shown.

Reply:
I see, lemme try that